src/main/cpp/apple_fsnotifier.cpp

```diff
@@ -1,116 +1,145 @@
-/*
- * Initial version copied from https://github.com/JetBrains/intellij-community/blob/master/native/fsNotifier/mac/fsnotifier.c
- */
-// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
-
-// Important notes on FSEvents: https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/FSEvents_ProgGuide/UsingtheFSEventsFramework/UsingtheFSEventsFramework.html
-
-/*
- * Apple specific functions.
- */
 #if defined(__APPLE__)
 
 #include "net_rubygrapefruit_platform_internal_jni_OsxFileEventFunctions.h"
 #include "generic.h"
 #include <CoreServices/CoreServices.h>
 #include <pthread.h>
-#include <strings.h>
 
-bool invalidStateDetected = false;
+class Server;
+
+static void handleEventsCallback(
+    ConstFSEventStreamRef streamRef,
+    void *clientCallBackInfo,
+    size_t numEvents,
+    void *eventPaths,
+    const FSEventStreamEventFlags eventFlags[],
+    const FSEventStreamEventId eventIds[]);
+
+static void *EventProcessingThread(void *data);
+
+class Server {
+public:
+    Server(JavaVM *jvm, JNIEnv *env, jobject watcherCallback, CFMutableArrayRef rootsToWatch, long latencyInMillis);
+    ~Server();
+
+private:
+    void run();
+    friend void *EventProcessingThread(void *data);
+
+    void handleEvents(
+        size_t numEvents,
+        char **eventPaths,
+        const FSEventStreamEventFlags eventFlags[],
+        const FSEventStreamEventId eventIds[]);
+    friend void handleEventsCallback(
+        ConstFSEventStreamRef streamRef,
+        void *clientCallBackInfo,
+        size_t numEvents,
+        void *eventPaths,
+        const FSEventStreamEventFlags eventFlags[],
+        const FSEventStreamEventId eventIds[]);
+
+    void handleEvent(JNIEnv *env, char* path, FSEventStreamEventFlags flags);
+
+    // TODO: Move this to somewhere else
+    JNIEnv* getThreadEnv();
 
-typedef struct watch_details {
+    JavaVM *jvm;
+    jobject watcherCallback;
     CFMutableArrayRef rootsToWatch;
     FSEventStreamRef watcherStream;
     pthread_t watcherThread;
-    JavaVM *jvm;
-    JNIEnv *env;
-    jobject watcherCallback;
     CFRunLoopRef threadLoop;
-} watch_details_t;
+    bool invalidStateDetected;
+};
+
+Server::Server(JavaVM *jvm, JNIEnv *env, jobject watcherCallback, CFMutableArrayRef rootsToWatch, long latencyInMillis) {
+    this->jvm = jvm;
+    this->watcherCallback = env->NewGlobalRef(watcherCallback);
```

Comment:
Do we need to check here if `this->watcherCallback` is `NULL`?

Reply:
I'll add a note to check this as well once we add proper error handling.